com.hazelcast.internal.serialization.impl.compact.integration.CompactFormatSplitBrainTest [API-1760]

_master_ (commit 40701e7d06bfd9feff165c04fd0744f2ff365f29)

Failed on Sonar build (Oracle JDK 11): https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar/1353/testReport/com.hazelcast.internal.serialization.impl.compact.integration/CompactFormatSplitBrainTest/testSplitBrainHealing_whenBothClusterHaveSameSchemas/

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: Cluster size is not correct. Expected: 2, actual: 1, instance index: 0
	at org.junit.Assert.fail(Assert.java:89)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:1021)
	at com.hazelcast.test.HazelcastTestSupport.lambda$assertClusterSizeEventually$9(HazelcastTestSupport.java:1056)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1236)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1253)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1056)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1034)
	at com.hazelcast.internal.serialization.impl.compact.integration.CompactFormatSplitBrainTest.splitCluster(CompactFormatSplitBrainTest.java:171)
	at com.hazelcast.internal.serialization.impl.compact.integration.CompactFormatSplitBrainTest.testSplitBrainHealing_whenBothClusterHaveSameSchemas(CompactFormatSplitBrainTest.java:120)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

</details>

Standard output can be found here - https://console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-sonar/1353/

--- a/hazelcast/src/test/java/com/hazelcast/client/ClientSplitBrainTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/ClientSplitBrainTest.java
@@ -220,8 +220,7 @@ public class ClientSplitBrainTest extends ClientTestSupport {
 
     @Test
     public void testMemberList_afterConnectingToOtherHalf() {
-        Config config = smallInstanceConfig();
-        config.getJetConfig().setEnabled(false);
+        Config config = smallInstanceConfigWithoutJetAndMetrics();
         HazelcastInstance instance1 = factory.newHazelcastInstance(config);
 
         // Client now has the cluster view listener registered to instance1

--- a/hazelcast/src/test/java/com/hazelcast/client/ClientSplitBrainTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/ClientSplitBrainTest.java
@@ -239,12 +239,9 @@ public class ClientSplitBrainTest extends ClientTestSupport {
         blockCommunicationBetween(instance2, instance4);
 
         // make sure that each member quickly drops the other from their member list
-        suspectMember(instance1, instance2);
-        suspectMember(instance2, instance1);
-        suspectMember(instance3, instance2);
-        suspectMember(instance2, instance3);
-        suspectMember(instance4, instance2);
-        suspectMember(instance2, instance4);
+        closeConnectionBetween(instance1, instance2);
+        closeConnectionBetween(instance3, instance2);
+        closeConnectionBetween(instance4, instance2);
 
         assertClusterSizeEventually(3, instance1, instance3, instance4);
         assertClusterSizeEventually(1, instance2);

--- a/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/ClientCompactSplitBrainTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/ClientCompactSplitBrainTest.java
@@ -68,7 +68,7 @@ public class ClientCompactSplitBrainTest extends ClientTestSupport {
 
     @Test
     public void testLocalSchemasAreSent_whenClientReconnectsToOtherHalf() {
-        Config config = smallInstanceConfig();
+        Config config = smallInstanceConfigWithoutJetAndMetrics();
 
         HazelcastInstance instance1 = factory.newHazelcastInstance(config);
         HazelcastInstance instance2 = factory.newHazelcastInstance(config);
@@ -79,8 +79,7 @@ public class ClientCompactSplitBrainTest extends ClientTestSupport {
         blockCommunicationBetween(instance1, instance2);
 
         // make sure that each member quickly drops the other from their member list
-        suspectMember(instance1, instance2);
-        suspectMember(instance2, instance1);
+        closeConnectionBetween(instance1, instance2);
 
         assertClusterSizeEventually(1, instance1);
         assertClusterSizeEventually(1, instance2);
@@ -122,8 +121,7 @@ public class ClientCompactSplitBrainTest extends ClientTestSupport {
 
     @Test
     public void testSchemaReplicationRetried_whenClientIsConnectedToBothHalvesOfTheSplit() {
-        Config config = smallInstanceConfig();
-        config.getJetConfig().setEnabled(false);
+        Config config = smallInstanceConfigWithoutJetAndMetrics();
 
         HazelcastInstance instance1 = factory.newHazelcastInstance(config);
         HazelcastInstance instance2 = factory.newHazelcastInstance(config);

--- a/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/CompactFormatSplitBrainTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/CompactFormatSplitBrainTest.java
@@ -205,6 +205,7 @@ public class CompactFormatSplitBrainTest extends HazelcastTestSupport {
         // make sure that cluster is split
         assertClusterSizeEventually(largerBrain.instances.size(), largerBrain.instances);
         assertClusterSizeEventually(smallerBrain.instances.size(), smallerBrain.instances);
+        waitAllForSafeState(allInstances);
     }
 
     private void mergeCluster() {

--- a/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/CompactFormatSplitBrainTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/integration/CompactFormatSplitBrainTest.java
@@ -25,7 +25,6 @@ import com.hazelcast.spi.merge.PutIfAbsentMergePolicy;
 import com.hazelcast.spi.properties.ClusterProperty;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
-import com.hazelcast.test.SplitBrainTestSupport;
 import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
 import example.serialization.EmployeeDTO;
@@ -36,11 +35,14 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collection;
 import java.util.List;
+import java.util.Random;
 
 import static com.hazelcast.internal.serialization.impl.compact.CompactTestUtil.assertSchemasAvailable;
+import static com.hazelcast.test.SplitBrainTestSupport.blockCommunicationBetween;
+import static com.hazelcast.test.SplitBrainTestSupport.unblockCommunicationBetween;
 import static org.junit.Assert.assertEquals;
 
 @RunWith(HazelcastParallelClassRunner.class)
@@ -49,27 +51,30 @@ public class CompactFormatSplitBrainTest extends HazelcastTestSupport {
 
     private final TestHazelcastFactory factory = new TestHazelcastFactory();
 
-    private HazelcastInstance instance1;
-    private HazelcastInstance instance2;
-    private HazelcastInstance instance3;
-    private HazelcastInstance instance4;
-    private HazelcastInstance instance5;
-
-    private List<HazelcastInstance> splitA;
-    private List<HazelcastInstance> splitB;
-
+    private Brain largerBrain;
+    private Brain smallerBrain;
+    private List<HazelcastInstance> allInstances;
 
     @Before
     public void setUp() {
         Config config = getMemberConfig();
-        instance1 = factory.newHazelcastInstance(config);
-        instance2 = factory.newHazelcastInstance(config);
-        instance3 = factory.newHazelcastInstance(config);
-        instance4 = factory.newHazelcastInstance(config);
-        instance5 = factory.newHazelcastInstance(config);
 
-        splitA = Arrays.asList(instance1, instance2, instance3);
-        splitB = Arrays.asList(instance4, instance5);
+        largerBrain = new Brain(Arrays.asList(
+                factory.newHazelcastInstance(config),
+                factory.newHazelcastInstance(config),
+                factory.newHazelcastInstance(config)
+        ));
+
+        smallerBrain = new Brain(Arrays.asList(
+                factory.newHazelcastInstance(config),
+                factory.newHazelcastInstance(config)
+        ));
+
+        allInstances = new ArrayList<>();
+        allInstances.addAll(largerBrain.instances);
+        allInstances.addAll(smallerBrain.instances);
+
+        assertClusterSizeEventually(allInstances.size(), allInstances);
     }
 
     @After
@@ -81,108 +86,136 @@ public class CompactFormatSplitBrainTest extends HazelcastTestSupport {
     public void testSplitBrainHealing_whenSmallerClusterHasNoSchemas() {
         splitCluster();
 
-        // The data is only available in the larger cluster
-        fillEmployeeMapUsing(instance1);
-        fillNodeMapUsing(instance1);
+        // Schemas of the Employee and the Node classes are only available in the larger brain
+        fillEmployeeMapUsing(largerBrain.getRandomInstance());
+        fillNodeMapUsing(largerBrain.getRandomInstance());
 
         mergeCluster();
 
-        assertEmployeeMapSizeUsing(instance5);
-        assertNodeMapSizeUsing(instance5);
+        assertEmployeeMapSizeUsing(smallerBrain.getRandomInstance());
+        assertNodeMapSizeUsing(smallerBrain.getRandomInstance());
 
         assertSchemasAvailableInEveryMember();
 
-        assertQueryForEmployeeMapUsing(instance5);
-        assertQueryForNodeMapUsing(instance5);
+        assertQueryForEmployeeMapUsing(smallerBrain.getRandomInstance());
+        assertQueryForNodeMapUsing(smallerBrain.getRandomInstance());
     }
 
     @Test
     public void testSplitBrainHealing_whenLargeClusterHasNoSchemas() {
         splitCluster();
 
-        // The data is only available in the smaller cluster
-        fillEmployeeMapUsing(instance5);
-        fillNodeMapUsing(instance5);
+        // Schemas of the Employee and the Node classes are only available in the smaller brain
+        fillEmployeeMapUsing(smallerBrain.getRandomInstance());
+        fillNodeMapUsing(smallerBrain.getRandomInstance());
 
         mergeCluster();
 
-        assertEmployeeMapSizeUsing(instance1);
-        assertNodeMapSizeUsing(instance1);
+        assertEmployeeMapSizeUsing(largerBrain.getRandomInstance());
+        assertNodeMapSizeUsing(largerBrain.getRandomInstance());
 
         assertSchemasAvailableInEveryMember();
 
-        assertQueryForEmployeeMapUsing(instance1);
-        assertQueryForNodeMapUsing(instance1);
+        assertQueryForEmployeeMapUsing(largerBrain.getRandomInstance());
+        assertQueryForNodeMapUsing(largerBrain.getRandomInstance());
     }
 
     @Test
     public void testSplitBrainHealing_whenBothClusterHaveSameSchemas() {
         splitCluster();
 
-        // The data is available in both clusters
-        fillEmployeeMapUsing(instance1);
-        fillEmployeeMapUsing(instance5);
+        // Schemas of the Employee and the Node classes are only available in both of the brains
+        fillEmployeeMapUsing(largerBrain.getRandomInstance());
+        fillEmployeeMapUsing(smallerBrain.getRandomInstance());
 
-        fillNodeMapUsing(instance1);
-        fillNodeMapUsing(instance5);
+        fillNodeMapUsing(largerBrain.getRandomInstance());
+        fillNodeMapUsing(smallerBrain.getRandomInstance());
 
         mergeCluster();
 
-        assertEmployeeMapSizeUsing(instance1);
-        assertNodeMapSizeUsing(instance5);
+        assertEmployeeMapSizeUsing(largerBrain.getRandomInstance());
+        assertNodeMapSizeUsing(smallerBrain.getRandomInstance());
 
         assertSchemasAvailableInEveryMember();
 
-        assertQueryForEmployeeMapUsing(instance5);
-        assertQueryForNodeMapUsing(instance1);
+        assertQueryForEmployeeMapUsing(smallerBrain.getRandomInstance());
+        assertQueryForNodeMapUsing(largerBrain.getRandomInstance());
     }
 
     @Test
     public void testSplitBrainHealing_whenBothClusterHaveDifferentSchemas() {
         splitCluster();
 
-        // The data is only available in the large cluster
-        fillEmployeeMapUsing(instance1);
+        // The schema of the Employee class is only available in the larger brain
+        fillEmployeeMapUsing(largerBrain.getRandomInstance());
 
-        // Tha data is only available in the smaller cluster
-        fillNodeMapUsing(instance5);
+        // The schema of the Node class is only available in the smaller brain
+        fillNodeMapUsing(smallerBrain.getRandomInstance());
 
         mergeCluster();
 
-        assertEmployeeMapSizeUsing(instance5);
-        assertNodeMapSizeUsing(instance1);
+        assertEmployeeMapSizeUsing(smallerBrain.getRandomInstance());
+        assertNodeMapSizeUsing(largerBrain.getRandomInstance());
 
         assertSchemasAvailableInEveryMember();
 
-        assertQueryForEmployeeMapUsing(instance5);
-        assertQueryForNodeMapUsing(instance1);
+        assertQueryForEmployeeMapUsing(smallerBrain.getRandomInstance());
+        assertQueryForNodeMapUsing(largerBrain.getRandomInstance());
+    }
+
+    private static class Brain {
+        private final List<HazelcastInstance> instances;
+        private final Random random = new Random();
+
+        private Brain(List<HazelcastInstance> instances) {
+            this.instances = instances;
+        }
+
+        public HazelcastInstance getRandomInstance() {
+            return instances.get(random.nextInt(instances.size()));
+        }
+
+        public void splitFrom(Brain other) {
+            List<HazelcastInstance> otherBrainInstances = other.instances;
+            for (HazelcastInstance instance : instances) {
+                for (HazelcastInstance otherBrainInstance : otherBrainInstances) {
+                    blockCommunicationBetween(instance, otherBrainInstance);
+                }
+            }
+
+            for (HazelcastInstance instance : instances) {
+                for (HazelcastInstance otherBrainInstance : otherBrainInstances) {
+                    closeConnectionBetween(instance, otherBrainInstance);
+                }
+            }
+        }
+
+        public void mergeWith(Brain other) {
+            for (HazelcastInstance instance : instances) {
+                for (HazelcastInstance otherBrainInstance : other.instances) {
+                    unblockCommunicationBetween(instance, otherBrainInstance);
+                }
+            }
+        }
     }
 
     private void splitCluster() {
-        for (HazelcastInstance splitAInstance : splitA) {
-            for (HazelcastInstance splitBInstance : splitB) {
-                SplitBrainTestSupport.blockCommunicationBetween(splitAInstance, splitBInstance);
-                closeConnectionBetween(splitAInstance, splitBInstance);
-            }
-        }
+        largerBrain.splitFrom(smallerBrain);
 
-        // make sure that cluster is split as [1 , 2, 3] , [4, 5]
-        assertClusterSizeEventually(3, instance1, instance2, instance3);
-        assertClusterSizeEventually(2, instance4, instance5);
+        // make sure that cluster is split
+        assertClusterSizeEventually(largerBrain.instances.size(), largerBrain.instances);
+        assertClusterSizeEventually(smallerBrain.instances.size(), smallerBrain.instances);
     }
 
     private void mergeCluster() {
-        for (HazelcastInstance splitAInstance : splitA) {
-            for (HazelcastInstance splitBInstance : splitB) {
-                SplitBrainTestSupport.unblockCommunicationBetween(splitAInstance, splitBInstance);
-            }
-        }
+        largerBrain.mergeWith(smallerBrain);
 
-        assertClusterSizeEventually(5, instance1, instance2, instance3, instance4, instance5);
+        assertClusterSizeEventually(allInstances.size(), allInstances);
+        waitAllForSafeState(allInstances);
     }
 
     private Config getMemberConfig() {
-        Config config = smallInstanceConfig();
+        Config config = smallInstanceConfigWithoutJetAndMetrics();
         config.getMapConfig("employeeMap")
                 .getMergePolicyConfig()
                 .setPolicy(PutIfAbsentMergePolicy.class.getName());
@@ -195,14 +228,7 @@ public class CompactFormatSplitBrainTest extends HazelcastTestSupport {
     }
 
     private void assertSchemasAvailableInEveryMember() {
-        Collection<HazelcastInstance> instances = Arrays.asList(
-                instance1,
-                instance2,
-                instance3,
-                instance4,
-                instance5
-        );
-        assertSchemasAvailable(instances, EmployeeDTO.class, NodeDTO.class);
+        assertSchemasAvailable(allInstances, EmployeeDTO.class, NodeDTO.class);
     }
 
     private void fillEmployeeMapUsing(HazelcastInstance instance) {

